### PR TITLE
[Search 2.0] Add Search::Postgres::Tag behind a feature flag

### PR DESCRIPTION
**Pull request by rhymes**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [x] Feature
- [ ] Bug Fix
- [ ] Optimization
- [ ] Documentation Update

## Description

This adds "searching for tags" behind a feature flag, using the new PostgreSQL FTS added in #12905

A few considerations:

- Our current `/search/tags` endpoint, if it's not given a `name` parameter, [returns an arbitrary list of tags](https://dev.to/search/tags) as ES can both be used as a search engine but also to enumerate indexed documents. Given that nowhere in the frontend we use this latter capability, I decided not to implement it
- If a `name` parameter is not given on the URL, nothing happens and an empty result set is returned. I didn't change it to a 404 to avoid breaking the UI
- Thanks to @Ridhwana I decided to skip the recommended activation of feature flags by way of data update script as we don't want to have admins of other Forems to accidentally enable it until we've tested it thoroughly ourselves

## Related Tickets & Documents

RFC 0153: https://github.com/forem/rfcs/pull/153

Where we use this functionality in the frontend:

https://github.com/forem/forem/blob/4b902b01c039550b86d79298415808193e7f7ffc/app/javascript/shared/components/tags.jsx#L318-L319

## QA Instructions, Screenshots, Recordings

1. Open the website locally, go to http://localhost:3000/new or http://localhost:3000/listings/new and look for tags
2. Load the console, activate the flag with `FeatureFlag.enable(:search_2_tags)`
3. Repeat step 1

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## [Forem core team only] How will this change be communicated?

_Will this PR introduce a change that impacts Forem members or creators, the
development process, or any of our internal teams? If so, please note how you
will share this change with the people who need to know about it._

- [ ] I've updated the [Developer Docs](https://docs.forem.com) and/or
      [Admin Guide](https://forem.gitbook.io/forem-admin-guide/), or
      [Storybook](https://storybook.forem.com/) (for Crayons components)
- [ ] I've updated the README or added inline documentation
- [ ] I will share this change in a [Changelog](https://forem.dev/t/changelog)
      or in a [forem.dev](http://forem.dev) post
- [x] I will share this change internally with the appropriate teams
- [ ] I'm not sure how best to communicate this change and need help
- [ ] This change does not need to be communicated, and this is why not: _please
      replace this line with details on why this change doesn't need to be
      shared_

## [optional] Are there any post deployment tasks we need to perform?

We'll test it on selected Forems in production and monitor the performance in production

### CODE REVIEW

File: spec/requests/search_spec.rb [hunk 1]
```diff
@@ -1,26 +1,55 @@
...
+      # NOTE: [@rhymes] ES returns all tags if no parameter is given **but** this endpoint is only used by
+      # the `Tags` Preact component so there's no need for us to return all tags. After all it's a search endpoint,
+      # not an index page
+      it "returns nothing if there is no name parameter" do
+        get search_tags_path
+        expect(response.parsed_body["result"]).to be_empty
+      end
+
+      it "finds a tag by a partial name" do
+        tag = create(:tag, name: "elixir")
```
Comment (atsmith813) [inline spec/requests/search_spec.rb:47]: 😊

**Review by atsmith813 — APPROVED** [review]
Tested locally, works great 🚀 

**question (non-blocking)**

I saw in the PR description we have
> We'll test it on selected Forems in production and monitor the performance in production

What's our game plan for monitoring the performance of this new search?

File: app/models/tag.rb [hunk 2]
```diff
@@ -10,6 +10,11 @@ class Tag < ActsAsTaggableOn::Tag
   include Purgeable
   include Searchable
 
+  include PgSearch::Model
+  pg_search_scope :search_by_name,
```
Comment (citizen428) [inline app/models/tag.rb:14]: I know you want to measure performance in production, but pretty much every time I built a Postgres FTS I ended up adding `tsvector` columns:

https://github.com/Casecommons/pg_search#using-tsvector-columns

That said, the generated query seems fast enough in Blazer.

File: spec/services/search/postgres/tag_spec.rb [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+    end
+
+    it "order tags by decreasing hotness score" do
+      java = create(:tag, name: "java", hotness_score: 10)
+      javascript = create(:tag, name: "javascript", hotness_score: 20)
+
+      result = described_class.search_documents("jav")
+      tags = result.map { |r| r["name"] }
+
+      expect(tags).to match_array([javascript.name, java.name])
```
Comment (citizen428) [inline spec/services/search/postgres/tag_spec.rb:53]: `match_array` is an alternate form of `contain_exactly`, which does not take ordering into account.

[`match_array` docs](https://www.rubydoc.info/gems/rspec-expectations/RSpec%2FMatchers:match_array):

> An alternate form of `contain_exactly `that accepts the expected contents as a single array arg rather that splatted out as individual items.

```ruby
expect(results).to contain_exactly(1, 2)
# is identical to:
expect(results).to match_array([1, 2])
```

[`contain_exactly` docs](https://www.rubydoc.info/gems/rspec-expectations/RSpec/Matchers#contain_exactly-instance_method), emphasis added:

> Passes if actual contains all of the expected **regardless of order**.

I threw together a [test spec](https://gist.github.com/citizen428/725ea91032f7c270158afaf667d573d3) to illustrate this.

Since this spec is specifically about ordering we should use a different matcher instead;

```suggestion
      expect(tags).to eq([javascript.name, java.name])
```

Comment (rhymes) [reply spec/services/search/postgres/tag_spec.rb:53]: This stuff is why I dislike RSpec :D

**Comment by citizen428** [comment]
> Checking the query plan for this on DEV in production, it looks like tag search happens in O(n) time because we don't have any indexes on it yet. Even so, it still completes in 75ms at the server, so we should be fine to try it before optimizing.

FWIW, I tried the same via Blazer yesterday and reached the same conclusion:

> That said, the generated query seems fast enough in Blazer.

> I'm not sure yet how to index for this since we're allowing partial words,

B-tree indexes on strings work for left anchored `LIKE`-searches. For FTS you can do trigram indexes (we already have the extension enabled) and use the right search function (`to_tsquery` which the gem takes care of for us).

**Comment by jgaskins** [comment]
@citizen428 I meant which expression to index on. :-) Storing it in a column would return it as an attribute of a `Tag` in ActiveRecord, but it's not useful for that. Since we only need it to make the query faster, we can instead index the expression. What I normally do for FTS in Postgres is something like this:

```sql
CREATE INDEX CONCURRENTLY ON tags USING GIN (to_tsvector('simple', name))
```

But the query isn't constructing `tsvector`s based purely on the `name` column and we also need to index on `ts_rank` so ordering will be faster, as well. That's what I meant by how to index for it. :-) I couldn't find anything useful for it in the gem's docs — it seemed to be all about what it calls "multisearch", which we don't appear to be using.

**Comment by citizen428** [comment]
@jgaskins

> I meant which expression to index on.

Ah, I see. 

> Storing it in a column would return it as an attribute of a Tag in ActiveRecord, but it's not useful for that. 

Generally in places where it matters we explicitly `select` attributes anyway, so I don't feel strongly about avoiding that column. I usually just follow the approach outlined here: https://thoughtbot.com/blog/optimizing-full-text-search-with-postgres-tsvector-columns-and-triggers

>  I couldn't find anything useful for it in the gem's docs — it seemed to be all about what it calls "multisearch", which we don't appear to be using.

Everything the docs mention for multisearch pretty much applies to single model search because they are implemented the same way (multisearch is a single table search against a table of search documents).

**Comment by jgaskins** [comment]
@citizen428 

> Generally in places where it matters we explicitly `select` attributes anyway, so I don't feel strongly about avoiding that column.

It's not just about reducing the row width, although that's certainly a part of it. A couple more items to consider:

1. the `pg` gem doesn't know how to decode a `tsvector` and delivers a warning about deserializing it as a string:

   ```
   [1] pry(main)> Tag.select("to_tsquery(name) AS tsquery").limit(1).first.tsquery
     Tag Load (1.5ms)  SELECT to_tsquery(name) AS tsquery FROM "tags" ORDER BY "tags"."id" ASC LIMIT $1  [["LIMIT", 1]]
   unknown OID 3615: failed to recognize type of 'tsquery'. It will be treated as String.
   => "'beginn'"
   ```

2. Storing it inside the table requires changing the schema dump format:

   Since we can’t dump a tsvector column to schema.rb, we need to switch to the SQL schema format in our config/application.rb

If it were necessary to store the `tsvector` in the table itself in order to optimize FTS, these are all things we could live with. But since it's not, there's no reason to. We can instead add the indexes we need and nobody has to work around these things.

**Comment by rhymes** [comment]
@jgaskins @citizen428 

Re-reading the query plan I noticed we don't have an index on `tags.supported`, I think that will help and it has little cost. We can even create a partial index. DEV has 268 supported tags and 82k non supported tags 👀 

An optimization technique I explored locally was to build JSON directly with `array_to_json(array_agg(row_to_json(t)))` but I don't think this kind of micro optimization is worth it yet. We should gather some data before tuning these things. The index on the `WHERE` condition is definitely more straightforward and doesn't complicate the code.

On the matter of indexing FTS and optimization on `tsvector` or trigrams I think we can go down that road when we need to. We'll need to enable search on much larger result sets (ie. Users, which is 580K+), that might require all those optimizations.

I'm going to increase the number of tags locally and report back some data or the before and after adding the index

**Comment by rhymes** [comment]
So, I did a few tests. I created 82k+ tags locally and 270 of them set as supported. These are my query plans:

<details>
<summary>No index</summary>

```sql
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                        |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Sort  (cost=2308.24..2308.24 rows=1 width=102) (actual time=15.437..15.439 rows=0 loops=1)                                                                        |
|   Sort Key: (ts_rank(to_tsvector('simple'::regconfig, COALESCE((tags_1.name)::text, ''::text)), '''ruby'':*'::tsquery, 0)) DESC, tags.id, tags.hotness_score DESC |
|   Sort Method: quicksort  Memory: 25kB                                                                                                                            |
|   ->  Nested Loop  (cost=0.29..2308.23 rows=1 width=102) (actual time=15.402..15.403 rows=0 loops=1)                                                              |
|         ->  Seq Scan on tags  (cost=0.00..2188.09 rows=14 width=98) (actual time=0.009..14.824 rows=268 loops=1)                                                  |
|               Filter: supported                                                                                                                                   |
|               Rows Removed by Filter: 81894                                                                                                                       |
|         ->  Index Scan using tags_pkey on tags tags_1  (cost=0.29..8.56 rows=1 width=29) (actual time=0.002..0.002 rows=0 loops=268)                              |
|               Index Cond: (id = tags.id)                                                                                                                          |
|               Filter: (to_tsvector('simple'::regconfig, COALESCE((name)::text, ''::text)) @@ '''ruby'':*'::tsquery)                                               |
|               Rows Removed by Filter: 1                                                                                                                           |
| Planning Time: 2.291 ms                                                                                                                                           |
| Execution Time: 15.484 ms                                                                                                                                         |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```

<details>
<summary>Full index</summary>

```sql
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                        |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Sort  (cost=128.79..128.79 rows=1 width=102) (actual time=0.612..0.613 rows=0 loops=1)                                                                            |
|   Sort Key: (ts_rank(to_tsvector('simple'::regconfig, COALESCE((tags_1.name)::text, ''::text)), '''ruby'':*'::tsquery, 0)) DESC, tags.id, tags.hotness_score DESC |
|   Sort Method: quicksort  Memory: 25kB                                                                                                                            |
|   ->  Nested Loop  (cost=0.58..128.78 rows=1 width=102) (actual time=0.608..0.609 rows=0 loops=1)                                                                 |
|         ->  Index Scan using tags_supported on tags  (cost=0.29..8.64 rows=14 width=98) (actual time=0.012..0.056 rows=268 loops=1)                               |
|               Index Cond: (supported = true)                                                                                                                      |
|         ->  Index Scan using tags_pkey on tags tags_1  (cost=0.29..8.56 rows=1 width=29) (actual time=0.002..0.002 rows=0 loops=268)                              |
|               Index Cond: (id = tags.id)                                                                                                                          |
|               Filter: (to_tsvector('simple'::regconfig, COALESCE((name)::text, ''::text)) @@ '''ruby'':*'::tsquery)                                               |
|               Rows Removed by Filter: 1                                                                                                                           |
| Planning Time: 3.104 ms                                                                                                                                           |
| Execution Time: 0.635 ms                                                                                                                                          |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
</details>

<details>
<summary>Partial index</summary>

```sql
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
| QUERY PLAN                                                                                                                                                        |
|-------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| Sort  (cost=132.54..132.54 rows=1 width=102) (actual time=0.586..0.587 rows=0 loops=1)                                                                            |
|   Sort Key: (ts_rank(to_tsvector('simple'::regconfig, COALESCE((tags_1.name)::text, ''::text)), '''ruby'':*'::tsquery, 0)) DESC, tags.id, tags.hotness_score DESC |
|   Sort Method: quicksort  Memory: 25kB                                                                                                                            |
|   ->  Nested Loop  (cost=0.42..132.53 rows=1 width=102) (actual time=0.582..0.583 rows=0 loops=1)                                                                 |
|         ->  Index Scan using tags_supported on tags  (cost=0.13..12.39 rows=14 width=98) (actual time=0.003..0.047 rows=268 loops=1)                              |
|         ->  Index Scan using tags_pkey on tags tags_1  (cost=0.29..8.56 rows=1 width=29) (actual time=0.002..0.002 rows=0 loops=268)                              |
|               Index Cond: (id = tags.id)                                                                                                                          |
|               Filter: (to_tsvector('simple'::regconfig, COALESCE((name)::text, ''::text)) @@ '''ruby'':*'::tsquery)                                               |
|               Rows Removed by Filter: 1                                                                                                                           |
| Planning Time: 0.338 ms                                                                                                                                           |
| Execution Time: 0.612 ms                                                                                                                                          |
+-------------------------------------------------------------------------------------------------------------------------------------------------------------------+
```
</summary>

I think a 24x improvement in the query plan should be enough to delay any further improvement on the query on this specific table.

File: spec/services/search/postgres/tag_spec.rb [hunk 4]
```diff
@@ -0,0 +1,64 @@
+require "rails_helper"
+
+RSpec.describe Search::Postgres::Tag, type: :service do
```
Comment (maestromac) [inline spec/services/search/postgres/tag_spec.rb:3]: I appreciate the lack of `let(:tag)` here. :relieved:

**Comment by citizen428** [comment]
@jgaskins 

tl;dr: Feel free to ignore my ramblings, they are moot in the context of tags FTS which is more than fast enough. 

> The pg gem doesn't know how to decode a tsvector and delivers a warning about deserializing it as a string:

I know.  In the past, I still sometimes found it useful to be able to inspect `tsvector` columns from a Rails console when debugging.
 ```ruby
Article.select("to_tsvector('the man ran to the other man') AS tsvector").limit(1).first.tsvector
"'man':2,7 'ran':3"
```

Let me summarize my current understanding.

1. `tsvector`: the lexemes of the content, stemmed, sorted, and deduplicated:
    ```sql
    Forem_development=# SELECT 'a fat cat sat on a mat and ate a fat rat'::tsvector;
                          tsvector
    ----------------------------------------------------
     'a' 'and' 'ate' 'cat' 'fat' 'mat' 'on' 'rat' 'sat'
    ```
2. `tsquery`: the lexemes to be searched for. You can't turn arbitrary text into a `tsquery`, that will result in a syntax error:
    ```sql
    Forem_development=# SELECT to_tsquery('a fat cat sat on a mat and ate a fat rat');
    ERROR:  syntax error in tsquery: "a fat cat sat on a mat and ate a fat rat"
    ```
   Instead, you have to use an actual query, with search operators like `&` or `|`.

I assume you know all of this, I only went into this much detail to ensure we are on the same page. I also got a bit confused, because your index example used `to_tsvector`, and your `pg` example `to_tsquery`.

Anyway, I'm by no means an indexing export for FTS, but I genuinely want to understand the tradeoffs between a dedicated `tsvector` column and an index better.

This [SO answer](https://stackoverflow.com/a/59763009/220147) has some interesting points I wasn't aware of for example:

> The downside of not storing the tsvector is that it will have to be recompute the tsvector from the raw text in order to "recheck" that the row meets the query. This can be very slow.
> 
> Rechecks are necessary if the size of the bitmap of candidate matches overflows work_mem. For some operators rechecks are always required, such as the phrase match operators <->, <2>, etc.

This eventually led me to a [similar answer on the DBA StackExchange](https://dba.stackexchange.com/a/251362/141), which offers an interesting indexing alternative:

> If you use proximity searching (like "phraseto_tsquery" for example), with a functional index it will have to re-parse each match candidate document to a tsvector and check it for the right sequence and spacing of words. This can be quite slow, especially if the number of candidates is much higher than the number of final results. If the tsvector is stored, it can just read it and not reparse the document, this is much faster. I think other advanced features like "ts_headline" may be in the same situation--but I haven't tested them.
> 
> Even if you just use "@@", I think that if the bitmap of the number of results doesn't fit in "work_mem", then it will also need to reparse the documents in order to recheck candidate matching blocks which "overflowed". Of course in that case increasing "work_mem" would be likely be a better option than adding a column.
> 
> For what it is worth, if you use RUM rather than GIN, it will solve the the problem with phraseto_tsquery on functional indexes.

(Note: we probably wouldn't be using this specific search function, but maybe the potentially relevant `websearch_to_tsquery` behaves similarly, I don't know.)

I wasn't aware of RUM indexing, but [this post](https://medium.datadriveninvestor.com/the-rum-index-and-full-text-search-using-postgresql-1c08fb3bf540) on Medium provided some clarity.

> Storing it inside the table requires changing the schema dump format:

For me, that's a plus. 😃 I honestly don't like remembering what can and what cannot be serialized in Rails' schema DSL (or being constrained by it), so I prefer `structure.sql`.

**Review by citizen428 — APPROVED** [review]

**Comment by rhymes** [comment]
@citizen428 thanks for the links, I wasn't aware of RUM indexes either, I see it's an external plugin.

We're definitely going to use `ts_headline` for search results, as we need highligthing. We'll test performance and query plans when we get to those and decide accordingly (maybe we'll need a `tsvector` column on bigger tables).
